### PR TITLE
cluster: validate topo changes after edit-config

### PR DESCRIPTION
**Pull request by AstroProfundis**
### What problem does this PR solve? 
add validation for topology diff to forbid changes of immutable fields

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: pkg/cluster/edit/diff.go [hunk 1]
```diff
@@ -16,10 +16,19 @@ package edit
...
 	"fmt"
 	"io"
+	"strconv"
+	"strings"
 
+	"github.com/r3labs/diff"
 	"github.com/sergi/go-diff/diffmatchpatch"
 )
 
+const (
```
Comment (lonng) [inline pkg/cluster/edit/diff.go:26]: Prefer to move this into a common package, which can be used by `dm/cluster`.

File: pkg/cluster/edit/diff.go [hunk 2]
```diff
@@ -29,3 +38,62 @@ func ShowDiff(t1 string, t2 string, w io.Writer) {
...
+			if c.Type == diff.UPDATE && c.Path[len(c.Path)-1] == validateTagEditable {
+				// If the field is marked as editable, it is allowed to be modified no matter
+				// its parent level element is marked as editable or not
+				continue
+			}
+
+			pathEditable := true
+			pathIgnore := false
+			for _, p := range c.Path {
+				if _, err := strconv.Atoi(p); err == nil {
```
Comment (lonng) [inline pkg/cluster/edit/diff.go:73]: I think we cannot ignore the slice, for example, user add a new host into `tidb_servers`.

File: pkg/cluster/edit/diff.go [hunk 3]
```diff
@@ -29,3 +38,62 @@ func ShowDiff(t1 string, t2 string, w io.Writer) {
...
+				}
+			}
+			// if the path has any ignorable item, just ignore it
+			if pathIgnore || pathEditable && (c.Type == diff.CREATE || c.Type == diff.DELETE) {
+				// If *every* parent elements on the path are all marked as editable,
+				// AND the field itself is marked as editable, it is allowed to add or delete
+				continue
+			}
+		}
+		msg = append(msg, fmt.Sprintf("(%s) %s '%v' -> '%v'", c.Type, strings.Join(c.Path, "."), c.From, c.To))
```
Comment (lonng) [inline pkg/cluster/edit/diff.go:92]: I think we should generate more readable error message. e.g:

1. update: `cannot change the xxx field from 1 to 2`
2. delete: `cannot delete the field xxx`
3. create: `cannot add the field xxx into yyy`

File: pkg/cluster/edit/diff.go (hunk 2 above)
Comment (AstroProfundis) [reply pkg/cluster/edit/diff.go:73]: `tidb_servers` is not ignored, `server_configs.tidb` is.

**Pull request merged**